spec.html

```diff
@@ -36075,6 +36075,16 @@ <h1>EnterCriticalSection ( _WL_ )</h1>
         <emu-alg>
           1. Assert: The calling agent is not in the critical section for any WaiterList.
           1. Wait until no agent is in the critical section for _WL_, then enter the critical section for _WL_ (without allowing any other agent to enter).
+          1. If there was an agent in the critical section for _WL_ prior to entry, then
+            1. Let _execution_ be the [[CandidateExecution]] field of the surrounding agent's Agent Record.
+            1. Let _eventsRecord_ be the Agent Events Record in _execution_.[[EventsRecords]] whose [[AgentSignifier]] is AgentSignifier().
+            1. Let _entererEventList_ be _eventsRecord_.[[EventList]].
+            1. Let _L_ be the Agent Signifier of the agent in the critical section immediately preceding the surrounding agent's entry.
+            1. Let _leaverEventList_ be the [[EventList]] field of the element in _execution_.[[EventRecords]] whose [[AgentSignifier]] is _L_.
+            1. Let _enterEvent_ and _leaveEvent_ be new Synchronize events.
+            1. Append _enterEvent_ to _entererEventList_.
+            1. Append _leaveEvent_ to _leaverEventList_.
```

Comment:
Morally, this will create a synchronization in the wrong place.

The leaver's (unpaired) synchronization event should be inserted into the event list at the moment of leaving the critical section, and with subsequent entries into the critical section creating the paired synchronization event. Otherwise, you allow executions where the leaver's synchronization event shows up in the agent-order _after_ other events which are program-order _before_ it (i.e. the notifying thread could have continued execution before the waiter gets a chance to re-enter the critical section). This is potentially un-observable, but it's un-intuitive and might cause inconsistencies later on if other features which expose operational interleavings are added.

I'm also a little iffy about the prose identification of the "immediately preceding" agent. This seems like the sort of thing that should be explicitly tracked as state in a semantic object.

The minimum patch would be to make the WaiterList keep a record of the last exiting synchronization event, updated with every LeaveCriticalSection, and synchronized on with every EnterCriticalSection.

Reply:
Thank you as always, @conrad-watt. These are all good suggestions, and I will try to incorporate them.